onnxruntime/core/optimizer/compute_optimizer/upstream_reshape_actors.cc

```diff
@@ -0,0 +1,283 @@
+                                 ": reshape's data input rank < passthrough node's input rank");
+      return false;
+    }
+
+    if (ret.size() < 2) {
+      LOG_DEBUG_INFO(logger, "Fail SimplePointwiseReshapeActor::PreCheck for node " + current_node.Name() +
+                                 ": full broadcasted shape has dim < 2.");
+      return false;
+    }
+
+    all_input_cmp_rets[input_idx] = ret;
+
+    if (ret[0] == DimCompareRet::NotExist && ret[1] == DimCompareRet::NotExist) {
+      // Don't need to propagate Reshape since the two leading dims to flatten do not ext.
+      LOG_DEBUG_INFO(logger, "In SimplePointwiseReshapeActor::PreCheck for node " + current_node.Name() +
+                                 ": skip propagating for the input because the merged dims does not exist.");
+      continue;
+    } else if (ret[0] == DimCompareRet::Equal && ret[1] == DimCompareRet::Equal) {
+      propagate_input_indices.push_back(input_idx);
+    } else {
+      LOG_DEBUG_INFO(logger, "Fail SimplePointwiseReshapeActor::PreCheck for node " + current_node.Name() +
+                                 ": not supported cases for two leading dims check.");
+      return false;
+    }
+
+    // All other dims should be equal
+    for (int dim_index = 2; dim_index < static_cast<int>(ret.size()); ++dim_index) {
+      if (ret[dim_index] != DimCompareRet::Equal) {
+        LOG_DEBUG_INFO(logger, "Fail SimplePointwiseReshapeActor::PreCheck for node " + current_node.Name() +
+                                   ": unflatten dims should all be equal to full broadcast shape.");
+        return false;
+      }
+    }
+  }
+
+  if (AreAllOutputShapesEqual) {
+    // Make sure once Reshape is moved before the target node, all its outputs can be correctly reshaped.
+    std::unordered_map<int, int> output_indices;
+    for (size_t output_idx = 0; output_idx < current_node.OutputDefs().size(); ++output_idx) {
+      if (static_cast<int>(output_idx) == current_node_output_index) {
+        continue;
+      }
+
+      auto [success, out_cmp_ret] = CompareInputShapeWithOutputShape(data_input_arg->Shape(),
+                                                                     current_node.OutputDefs()[output_idx]->Shape());
+
+      if (!success) {
+        LOG_DEBUG_INFO(logger, "Fail SimplePointwiseReshapeActor::PreCheck for node " + current_node.Name() +
+                                   ": reshape's data input rank < passthrough node's output rank");
+        return false;
+      }
+
+      // All dims should be equal
+      for (int dim_index = 0; dim_index < static_cast<int>(out_cmp_ret.size()); ++dim_index) {
+        if (out_cmp_ret[dim_index] != DimCompareRet::Equal) {
+          LOG_DEBUG_INFO(logger, "Fail SimplePointwiseReshapeActor::PreCheck for node " + current_node.Name() +
+                                     ": output shapes not equal.");
+          return false;
+        }
+      }
+    }
+
+    shape_update_func = [&info](Node& node) -> void {
```

Comment:
You're assigning a local lambda to an external `shape_update_func`. Will this result in a segfault since the local lambda might no longer be a valid address when invoked externally?

Reply:
I think this should be fine, local lamda function can be passed to outside. There are few existing use cases for this. 

onnxruntime/core/providers/cpu/controlflow/scan_9.cc

```
  device_helpers_.transpose_func = [](const gsl::span<const size_t>& permutations, const Tensor& input,
                                      Tensor& output, Stream* /*no stream needed for cpu*/) -> Status {
    return TransposeBase::DoTranspose(permutations, input, output);
  };

  device_helpers_.set_data_to_zero_func = [](void* data, size_t size_in_bytes) -> Status {
    memset(data, 0, size_in_bytes);
    return Status::OK();
  };
```